File: ios/MenCare/Features/Staff/BlockSheet.swift
import SwiftUI
import Observation

@MainActor
@Observable
final class BlockViewModel {

    private let auth: AuthRepository
    private let blockRepo: TimeBlockRepository
    private let booking: BookingRepository
    private let catalog: CatalogRepository

    var reason: BlockReason = .permesso {
        didSet { scheduleConflictCheck() }
    }

    private(set) var date: LocalDate = .today()
    private(set) var from = LocalTime(15, 30)
    private(set) var to = LocalTime(18, 0)
    var resolvedConflictIds: Set<String> = []
    var saved = false

    let state = LoadState()
    private(set) var colleagues: [Operator] = []
    /// Chi verrebbe travolto dal blocco: lo dice `GET /blocks/conflicts`, non
    /// piu' un confronto fatto in casa sull'agenda locale.
    private(set) var conflicts: [Appointment] = []
    private(set) var checkingConflicts = false
    var saveError: String?

    @ObservationIgnored private var operatorId: String?
    @ObservationIgnored private var conflictTask: Task<Void, Never>?

    init(auth: AuthRepository, blocks: TimeBlockRepository, booking: BookingRepository, catalog: CatalogRepository) {
        self.auth = auth
        self.blockRepo = blocks
        self.booking = booking
        self.catalog = catalog
    }

    func load() async {
        await state.run {
            operatorId = try await auth.currentUser()?.operatorId
            colleagues = try await catalog.operators().filter { $0.id != operatorId }
        }
        await checkConflicts()
    }

    private var totalMinutes: Int { to.minutesOfDay - from.minutesOfDay }

    var unresolvedConflicts: [Appointment] {
        conflicts.filter { !resolvedConflictIds.contains($0.id) }
    }

    var canSave: Bool { totalMinutes > 0 && unresolvedConflicts.isEmpty && !checkingConflicts }

    /// Ogni modifica della finestra rimanda la domanda al server: farla a ogni
    /// tocco del selettore vorrebbe dire una richiesta per minuto scorso.
    private func scheduleConflictCheck() {
        conflictTask?.cancel()
        conflictTask = Task {
            try? await Task.sleep(for: .milliseconds(250))
            guard !Task.isCancelled else { return }
            await checkConflicts()
        }
    }

    private func checkConflicts() async {
        guard let operatorId, totalMinutes > 0 else {
            conflicts = []
            return
        }
        checkingConflicts = true
        let candidate = TimeBlock(
            id: "", operatorId: operatorId, reason: reason, date: date, range: TimeRange(from, to)
        )
        conflicts = (try? await blockRepo.conflictsFor(candidate)) ?? []
        checkingConflicts = false
    }

    func setDate(_ newDate: LocalDate) {
        date = newDate
        resolvedConflictIds = []
        scheduleConflictCheck()
    }

    func setFrom(_ newFrom: LocalTime) {
        from = newFrom
        if newFrom >= to { to = newFrom.plusMinutes(60) }
        resolvedConflictIds = []
        scheduleConflictCheck()
    }

    func setTo(_ newTo: LocalTime) {
        to = newTo <= from ? from.plusMinutes(60) : newTo
        resolvedConflictIds = []
        scheduleConflictCheck()
    }

    func halfDay() { setRange(from: LocalTime(9, 0), to: LocalTime(13, 0)) }
    func fullDay() { setRange(from: LocalTime(9, 0), to: LocalTime(19, 0)) }

    private func setRange(from: LocalTime, to: LocalTime) {
        self.from = from
        self.to = to
        resolvedConflictIds = []
        scheduleConflictCheck()
    }

    /// Reassign a conflicting appointment to a colleague at the same time.
    func reassign(_ appointmentId: String, colleagueId: String) {
        guard let apt = conflicts.first(where: { $0.id == appointmentId }) else { return }
        Task {
            saveError = nil
            switch await booking.reschedule(appointmentId, newStart: apt.start, newOperatorId: colleagueId) {
            case .success: resolvedConflictIds.insert(appointmentId)
            case .failure(let error): saveError = error.displayMessage
            }
        }
    }

    /// "Proponi altro orario": Phase 1 marks the conflict handled — the client
    /// would get a reschedule proposal via push in Phase 2.
    func proposeNewTime(_ appointmentId: String) {
        resolvedConflictIds.insert(appointmentId)
    }

    func save() {
        guard let operatorId, canSave else { return }
        Task {
            saveError = nil
            let result = await blockRepo.createBlock(
                TimeBlock(
                    id: "", operatorId: operatorId, reason: reason, date: date,
                    range: TimeRange(from, to)
                )
            )
            switch result {
            case .success: saved = true
            case .failure(let error): saveError = error.displayMessage
            }
        }
    }
}

private enum Editing: Identifiable {
    case date, from, to
    var id: Self { self }
}

/// Ferie e permessi dell'operatore: la stessa tendina del titolare, senza la
/// scelta dell'operatore (è chi ha fatto l'accesso) e con i conflitti che si
/// risolvono qui.
struct BlockSheet: View {
    @Bindable var viewModel: BlockViewModel
    let onDismiss: () -> Void

    @State private var editing: Editing?
    @State private var headerHeight: CGFloat = 0
    @State private var formHeight: CGFloat = 0
    @State private var footerHeight: CGFloat = 0

    var body: some View {
        VStack(spacing: 0) {
            HStack {
                BarAction(text: L("block_cancel"), action: onDismiss, color: .ink)
                    .frame(width: 72, alignment: .leading)
                Text(L("block_title"))
                    .font(Typo.cormorant(21, weight: .regular))
                    .foregroundStyle(Color.ink)
                    .lineLimit(1)
                    .frame(maxWidth: .infinity)
                Color.clear.frame(width: 72, height: 1)
            }
            .padding(.horizontal, 20)
            .padding(.vertical, 6)
            .measureHeight($headerHeight)

            ScrollView {
                VStack(alignment: .leading, spacing: 18) {
                    BrandSectionLabel(text: L("block_reason"))
                    HStack(spacing: 8) {
                        ForEach(BlockReason.allCases, id: \.self) { reason in
                            BrandChip(
                                text: blockReasonLabel(reason),
                                selected: viewModel.reason == reason,
                                action: { viewModel.reason = reason },
                                fill: true
                            )
                        }
                    }

                    HStack(spacing: 9) {
                        PickerTile(label: L("block_date"), value: formatDateShort(viewModel.date)) { editing = .date }
                        PickerTile(label: L("block_from"), value: formatTime(viewModel.from)) { editing = .from }
                        PickerTile(label: L("block_to"), value: formatTime(viewModel.to)) { editing = .to }
                    }
                    HStack(spacing: 9) {
                        BrandChip(text: L("block_half_day"), selected: false, action: viewModel.halfDay, fill: true)
                        BrandChip(text: L("block_full_day"), selected: false, action: viewModel.fullDay, fill: true)
                    }

                    if !viewModel.unresolvedConflicts.isEmpty {
                        conflictsCard
                    }
                    FormErrorBanner(message: viewModel.saveError)
                }
                .padding(.horizontal, 20)
                .padding(.top, 20)
                .padding(.bottom, 12)
                .measureHeight($formHeight)
            }
            .scrollBounceBehavior(.basedOnSize)

            // Fuori dallo scroll: il bottone resta sempre a vista in fondo.
            AccentButton(
                text: L("block_cta", formatTime(viewModel.from), formatTime(viewModel.to)),
                action: viewModel.save,
                enabled: viewModel.canSave,
                height: 56,
                corner: 16
            )
            .padding(.horizontal, 20)
            .padding(.bottom, 12)
            .measureHeight($footerHeight)
        }
        .background(Color.bone)
        .task { await viewModel.load() }
        .fittedBrandSheet(height: headerHeight + formHeight + footerHeight)
        .onChange(of: viewModel.saved) {
            if viewModel.saved { onDismiss() }
        }
        .sheet(item: $editing) { mode in
            switch mode {
            case .date:
                BrandDatePickerDialog(
                    initial: viewModel.date,
                    onDismiss: { editing = nil },
                    onConfirm: { date in
                        viewModel.setDate(date)
                        editing = nil
                    }
                )
            case .from:
                BrandTimePickerDialog(
                    initial: viewModel.from,
                    onDismiss: { editing = nil },
                    onConfirm: { time in
                        viewModel.setFrom(time)
                        editing = nil
                    }
                )
            case .to:
                BrandTimePickerDialog(
                    initial: viewModel.to,
                    onDismiss: { editing = nil },
                    onConfirm: { time in
                        viewModel.setTo(time)
                        editing = nil
                    }
                )
            }
        }
    }

    private var conflictsCard: some View {
        AccentOutlinedCard {
            VStack(alignment: .leading, spacing: 0) {
                HStack(spacing: 10) {
                    RoundedRectangle(cornerRadius: 10)
                        .fill(Color.oliveWood)
                        .frame(width: 26, height: 26)
                        .overlay(Text("!").font(Typo.titleSmall).foregroundStyle(Color.bone))
                    Text(L("block_conflicts_count", viewModel.unresolvedConflicts.count))
                        .font(Typo.jost(15, weight: .medium))
                        .foregroundStyle(Color.ink)
                }
                ForEach(viewModel.unresolvedConflicts) { conflict in
                    Text(L(
                        "block_conflict_desc",
                        formatTime(conflict.time),
                        formatDurationLong(conflict.durationMinutes),
                        formatDateShort(conflict.date)
                    ))
                    .font(Typo.jost(12))
                    .foregroundStyle(Color.textMuted)
                    .padding(.top, 8)
                    HStack(spacing: 9) {
                        if let colleague = viewModel.colleagues.first {
                            Button {
                                viewModel.reassign(conflict.id, colleagueId: colleague.id)
                            } label: {
                                Text(L("block_reassign", String(colleague.name.split(separator: " ").first ?? "")))
                                    .font(Typo.titleSmall)
                                    .foregroundStyle(Color.bone)
                                    .padding(.horizontal, 14)
                                    .padding(.vertical, 11)
                                    .background(RoundedRectangle(cornerRadius: 10).fill(Color.ink))
                            }
                            .buttonStyle(.plain)
                        }
                        Button {
                            viewModel.proposeNewTime(conflict.id)
                        } label: {
                            Text(L("block_propose"))
                                .font(Typo.titleSmall)
                                .foregroundStyle(Color.ink)
                                .padding(.horizontal, 14)
                                .padding(.vertical, 11)
                                .background(RoundedRectangle(cornerRadius: 10).fill(Color.stone))
                        }
                        .buttonStyle(.plain)
                    }
                    .padding(.top, 10)
                }
            }
            .padding(14)
        }
    }
}

File: android/feature/staff/src/main/kotlin/com/devora/mencare/feature/staff/BlockSheet.kt
package com.devora.mencare.feature.staff

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.devora.mencare.core.common.formatDateShort
import com.devora.mencare.core.common.formatDurationLong
import com.devora.mencare.core.common.formatTime
import com.devora.mencare.core.designsystem.component.AccentButton
import com.devora.mencare.core.designsystem.component.BarAction
import com.devora.mencare.core.designsystem.component.BrandChip
import com.devora.mencare.core.designsystem.component.BrandDatePickerDialog
import com.devora.mencare.core.designsystem.component.BrandSectionLabel
import com.devora.mencare.core.designsystem.component.BrandTimePickerDialog
import com.devora.mencare.core.designsystem.component.PickerTile
import com.devora.mencare.core.designsystem.component.WarningCard
import com.devora.mencare.core.designsystem.theme.Bone
import com.devora.mencare.core.designsystem.theme.Cormorant
import com.devora.mencare.core.designsystem.theme.Ink
import com.devora.mencare.core.designsystem.theme.Radii
import com.devora.mencare.core.designsystem.theme.TextMuted
import com.devora.mencare.core.model.BlockReason

private enum class Editing { DATE, FROM, TO }

/**
 * Ferie e permessi dell'operatore: la stessa tendina del titolare, senza la
 * scelta dell'operatore (è chi ha fatto l'accesso) e con i conflitti che si
 * risolvono qui.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
internal fun BlockSheet(
    onDismiss: () -> Unit,
    viewModel: BlockViewModel = hiltViewModel(),
) {
    val state by viewModel.state.collectAsStateWithLifecycle()
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
    var editing by rememberSaveable { mutableStateOf<Editing?>(null) }

    // Il ViewModel vive con l'agenda: ogni apertura riparte da capo, come su iOS.
    val close = {
        viewModel.reset()
        onDismiss()
    }

    LaunchedEffect(state.saved) {
        if (state.saved) close()
    }

    ModalBottomSheet(
        onDismissRequest = close,
        sheetState = sheetState,
        containerColor = Bone,
        shape = RoundedCornerShape(topStart = 26.dp, topEnd = 26.dp),
    ) {
        Column(Modifier.navigationBarsPadding()) {
            Row(
                modifier = Modifier.fillMaxWidth().padding(horizontal = 20.dp, vertical = 6.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Box(Modifier.width(72.dp), contentAlignment = Alignment.CenterStart) {
                    BarAction(stringResource(R.string.block_cancel), close, color = Ink)
                }
                Text(
                    stringResource(R.string.block_title),
                    fontFamily = Cormorant,
                    fontSize = 21.sp,
                    color = Ink,
                    textAlign = TextAlign.Center,
                    maxLines = 1,
                    modifier = Modifier.weight(1f),
                )
                Spacer(Modifier.width(72.dp))
            }

            Column(
                modifier = Modifier
                    .weight(1f, fill = false)
                    .verticalScroll(rememberScrollState())
                    .padding(horizontal = 20.dp)
                    .padding(top = 20.dp, bottom = 12.dp),
                verticalArrangement = Arrangement.spacedBy(18.dp),
            ) {
                BrandSectionLabel(stringResource(R.string.block_reason))
                Row(horizontalArrangement = Arrangement.spacedBy(8.dp), modifier = Modifier.fillMaxWidth()) {
                    BlockReason.entries.forEach { reason ->
                        BrandChip(
                            text = stringResource(reasonLabel(reason)),
                            selected = state.reason == reason,
                            onClick = { viewModel.setReason(reason) },
                            modifier = Modifier.weight(1f),
                            fill = true,
                        )
                    }
                }

                Row(horizontalArrangement = Arrangement.spacedBy(9.dp), modifier = Modifier.fillMaxWidth()) {
                    PickerTile(
                        label = stringResource(R.string.block_date),
                        value = formatDateShort(state.date),
                        onClick = { editing = Editing.DATE },
                        modifier = Modifier.weight(1f),
                    )
                    PickerTile(
                        label = stringResource(R.string.block_from),
                        value = formatTime(state.from),
                        onClick = { editing = Editing.FROM },
                        modifier = Modifier.weight(1f),
                    )
                    PickerTile(
                        label = stringResource(R.string.block_to),
                        value = formatTime(state.to),
                        onClick = { editing = Editing.TO },
                        modifier = Modifier.weight(1f),
                    )
                }
                Row(horizontalArrangement = Arrangement.spacedBy(9.dp), modifier = Modifier.fillMaxWidth()) {
                    BrandChip(
                        text = stringResource(R.string.block_half_day),
                        selected = false,
                        onClick = viewModel::halfDay,
                        modifier = Modifier.weight(1f),
                        fill = true,
                    )
                    BrandChip(
                        text = stringResource(R.string.block_full_day),
                        selected = false,
                        onClick = viewModel::fullDay,
                        modifier = Modifier.weight(1f),
                        fill = true,
                    )
                }

                if (state.unresolvedConflicts.isNotEmpty()) {
                    ConflictsCard(state, viewModel)
                }
            }

            // Fuori dallo scroll: il bottone resta sempre a vista in fondo.
            AccentButton(
                text = stringResource(
                    R.string.block_cta,
                    formatTime(state.from),
                    formatTime(state.to),
                ),
                onClick = viewModel::save,
                enabled = state.canSave,
                height = 56.dp,
                shape = RoundedCornerShape(16.dp),
                modifier = Modifier.padding(horizontal = 20.dp).padding(bottom = 12.dp),
            )
        }
    }

    when (editing) {
        Editing.DATE -> BrandDatePickerDialog(
            initial = state.date,
            onDismiss = { editing = null },
            onConfirm = {
                viewModel.setDate(it)
                editing = null
            },
        )
        Editing.FROM -> BrandTimePickerDialog(
            initial = state.from,
            onDismiss = { editing = null },
            onConfirm = {
                viewModel.setFrom(it)
                editing = null
            },
        )
        Editing.TO -> BrandTimePickerDialog(
            initial = state.to,
            onDismiss = { editing = null },
            onConfirm = {
                viewModel.setTo(it)
                editing = null
            },
        )
        null -> Unit
    }
}

@Composable
private fun ConflictsCard(state: BlockUiState, viewModel: BlockViewModel) {
    val conflicts = state.unresolvedConflicts
    WarningCard(title = stringResource(R.string.block_conflicts_count, conflicts.size)) {
        Column {
            conflicts.forEach { conflict ->
                Spacer(Modifier.height(8.dp))
                Text(
                    stringResource(
                        R.string.block_conflict_desc,
                        formatTime(conflict.time),
                        formatDurationLong(conflict.durationMinutes),
                        formatDateShort(conflict.date),
                    ),
                    style = MaterialTheme.typography.bodySmall.copy(fontSize = 12.sp, lineHeight = 18.sp),
                    color = TextMuted,
                )
                Spacer(Modifier.height(10.dp))
                Row(horizontalArrangement = Arrangement.spacedBy(9.dp)) {
                    val colleague = state.colleagues.firstOrNull()
                    if (colleague != null) {
                        Text(
                            stringResource(R.string.block_reassign, colleague.name.substringBefore(' ')),
                            style = MaterialTheme.typography.titleSmall,
                            color = Bone,
                            modifier = Modifier
                                .clip(Radii.Sm)
                                .background(Ink)
                                .clickable { viewModel.reassign(conflict.id, colleague.id) }
                                .padding(horizontal = 14.dp, vertical = 11.dp),
                        )
                    }
                    Text(
                        stringResource(R.string.block_propose),
                        style = MaterialTheme.typography.titleSmall,
                        color = Ink,
                        modifier = Modifier
                            .clip(Radii.Sm)
                            .background(Bone)
                            .clickable { viewModel.proposeNewTime(conflict.id) }
                            .padding(horizontal = 14.dp, vertical = 11.dp),
                    )
                }
            }
        }
    }
}

private fun reasonLabel(reason: BlockReason): Int = when (reason) {
    BlockReason.PERMESSO -> R.string.block_reason_permesso
    BlockReason.PAUSA -> R.string.block_reason_pausa
    BlockReason.FERIE -> R.string.block_reason_ferie
    BlockReason.CORSO -> R.string.block_reason_corso
}
